pkg/instancetype/instancetype.go

```diff
@@ -521,6 +531,123 @@ func (m *methods) findClusterInstancetype(vm *virtv1.VirtualMachine) (*instancet
+	defaultName, defaultKind, err := m.inferDefaultsFromVolumes(vm, vm.Spec.Instancetype.InferFromVolume, DefaultInstancetypeAnnotation, DefaultInstancetypeKindAnnotation)
+	if err != nil {
+		return nil, err
+	}
+	return &v1.InstancetypeMatcher{
+		Name: defaultName,
+		Kind: defaultKind,
+	}, nil
+}
+
+func (m *methods) InferDefaultPreference(vm *virtv1.VirtualMachine) (*virtv1.PreferenceMatcher, error) {
+	if vm.Spec.Preference == nil || vm.Spec.Preference.InferFromVolume == "" {
+		return nil, nil
+	}
+	defaultName, defaultKind, err := m.inferDefaultsFromVolumes(vm, vm.Spec.Preference.InferFromVolume, DefaultPreferenceAnnotation, DefaultPreferenceKindAnnotation)
+	if err != nil {
+		return nil, err
+	}
+	return &v1.PreferenceMatcher{
+		Name: defaultName,
+		Kind: defaultKind,
+	}, nil
+}
+
+func (m *methods) inferDefaultsFromVolumes(vm *virtv1.VirtualMachine, inferFromVolumeName, defaultNameAnnotation, defaultKindAnnoation string) (string, string, error) {
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
+		if volume.Name != inferFromVolumeName {
+			continue
+		}
+		if volume.PersistentVolumeClaim != nil {
+			return m.inferDefaultsFromPVC(volume.PersistentVolumeClaim.ClaimName, vm.Namespace, defaultNameAnnotation, defaultKindAnnoation)
+		}
+		if volume.DataVolume != nil {
+			return m.inferDefaultsFromDataVolume(vm, volume.DataVolume.Name, defaultNameAnnotation, defaultKindAnnoation)
+		}
+		return "", "", fmt.Errorf("unable to infer defaults from Volume as type is not supported")
+	}
+	return "", "", fmt.Errorf("unable to find volume %s to infer defaults", inferFromVolumeName)
+}
+
+func inferDefaultsFromAnnotations(annotations map[string]string, defaultNameAnnotation, defaultKindAnnoation string) (string, string, error) {
+	defaultName, hasAnnotation := annotations[defaultNameAnnotation]
+	if !hasAnnotation {
+		return "", "", fmt.Errorf("unable to find required %s annotation on the volume", defaultNameAnnotation)
+	}
+	return defaultName, annotations[defaultKindAnnoation], nil
+}
+
+func (m *methods) inferDefaultsFromPVC(pvcName, pvcNamespace, defaultNameAnnotation, defaultKindAnnotation string) (string, string, error) {
+	pvc, err := m.clientset.CoreV1().PersistentVolumeClaims(pvcNamespace).Get(context.Background(), pvcName, metav1.GetOptions{})
+	if err != nil {
+		return "", "", err
+	}
+	return inferDefaultsFromAnnotations(pvc.Annotations, defaultNameAnnotation, defaultKindAnnotation)
+}
+
+func (m *methods) inferDefaultsFromDataVolume(vm *virtv1.VirtualMachine, dvName, defaultNameAnnotation, defaultKindAnnoation string) (string, string, error) {
+	if len(vm.Spec.DataVolumeTemplates) > 0 {
+		for _, dvt := range vm.Spec.DataVolumeTemplates {
+			if dvt.Name != dvName {
+				continue
+			}
+			return m.inferDefaultsFromDataVolumeSpec(dvt.Spec, defaultNameAnnotation, defaultKindAnnoation)
+		}
+	}
+	dv, err := m.clientset.CdiClient().CdiV1beta1().DataVolumes(vm.Namespace).Get(context.Background(), dvName, metav1.GetOptions{})
```

Comment:
What about garbage collected DVs?

Reply:
ACK apologies I thought I had handled that already.

Reply:
Makes me wonder what happens when a VM has specified a DV that was gargabe collected as volume? Will it fall back to PVC?

Reply:
Yes I believe we rely on the PVC having the same name as DataVolume that created it.